### DISCUSSION TITLE
External Contributor Troubleshooting

### DISCUSSION THREAD

**Discussion by conbrad** [bcgov/wps#2940] [General]
Hey @jmetev1 , I'm pulling your question over to discussions here since we'll lose track of the [PR discussion](https://github.com/bcgov/wps/pull/2935#issuecomment-1591962029) pretty quick.

> i'm also trying to run the whole app so i can do stuff requiring the backend. i've been trying to follow instructions using docker compose in root readme. but getting poetry errors. am i following the right instructions to start with? i also tried the instructions in Developing the application in a dev container, using vscode: but there is no such thing as docker-compose.vscode.yml

Yes, looks like those instructions are also out of date unfortunately. Dev containers are something I also haven't introduced to my workflow yet, but it looks like we have our dev container defined in the `./devcontainer` folder. I'll try and see if that's what we should be pointing at.

**Comment by jmetev1** [comment]
thank you. so right now people are generally just running api or other stuff independently without docker on their locals? i could probably figure that out if that's the norm.

**Reply by conbrad** [reply]
Yep, you'll just need:
- postgres locally on port 5432
- redis locally on port 6379
- api running via either
    - the `api` launch config command in vscode
    - `make run` in your terminal
- frontend running via either
    - yarn start in your terminal
    - the `Chrome` launch config command in vscode (I haven't used this in awhile either) 
    
Side note:
vscode just prompted me to open the project in a dev container upon startup and that seems to be building so far...

**Reply by conbrad** [reply]
<img width="466" alt="Screenshot 2023-06-14 at 4 35 28 PM" src="https://github.com/bcgov/wps/assets/1447136/56bfb2cb-5f15-4aaf-a973-5173a795f1ea">

**Reply by conbrad** [reply]
Also I guess this information should be more obvious in the READMEs... I can address that.

**Comment by jmetev1** [comment]
ha ok i changed my mind. everything i've ever done with python outside of docker has been pyenv mess. maybe i'll try and fix the docker compose setup instead. i already got past the poetry part.

**Reply by conbrad** [reply]
Fair enough! Let me know if you need a hand.

**Reply by conbrad** [reply]
> everything i've ever done with python outside of docker has been pyenv mess

There's an xkcd for that!
https://xkcd.com/1987/

For our setup, we have our `pyenv` pointing to python version 3.10.4, and then we use `poetry shell` to spin up a virtual environment using that. 

A few issues I've noticed recently is that `poetry` has started spinning up virtual environments in the project which causes some headache, I'll need to adjust `poetry.toml` to set `in-project = false` I believe. Another is that it complains `greenlet` is missing and `gdal` dependencies, which are a bit messy in that I'm not sure poetry can install these properly so they need to be installed after the other dependencies in `poetry install`. After that the API should be satisfied.

**Reply by jmetev1** [reply]
yeah that xkcd made me feel better. i feel like we are two non overlapping circles in a venn diagram. i don't know anything about dev containers and hate to mess running python stuff/redis/postgres directly. especially since i'm working on multiple projects at once i'm trying to keep things isolated. and you're more familiar with running the stuff directly and or dev container. 

i guess as a first step, i see api/dockerfile. and i see root dockerfile. and both seem setup for running the api. and both were updated 8 months ago. do you know the distinction?

**Reply by conbrad** [reply]
Yeah that makes sense. I'm on my work machine right now which only has this and a couple other projects so it's easier to manage, that's good feedback.

The root dockerfile is what's used to build and deploy on our infrastructure. The dockerfile in /api is what's used to build and run the api locally -- it'd be nice if we just could settle on one, but there's some infrastructure details in the root one that's required for our cluster. We may in fact be able to settle on 1 if we can run build and run the root dockerfile locally.

Note: we're also running M1 macs locally and have had some incompatibility issues with some of our OS dependencies so that's another reason why we've stopped using docker development locally (these issues may be resolved after some time though).

**Reply by jmetev1** [reply]
hmm interesting. i'm on m1 pro so do you expect that i would have same issues or is it something specific to you?

i might be getting close with docker compose. i'm failing on alembic part of this `bash -c "PYTHONPATH=. poetry run alembic upgrade head && poetry run uvicorn app.main:app --host 0.0.0.0 --reload --port 8080"`

i do need the alembic part right?

i tried to move forward with the react upgrade but it's breaking the cypress tests and i can't really debug without a working server. one thing i considered is that i can disable auth on the frontend and point the frontend to the prod servers. but of course you have auth on api requests also. no way i can get around that easily eh? i assume it's not super private data.

**Reply by jmetev1** [reply]
current alembic errors are:
wps-api-1 | File "/app/alembic/versions/d9c05cb16869_add_combustible_area_data_to_advisory_.py", line 47, in upgrade
wps-api-1 | with get_fuel_types_from_object_store() as fuel_types:
wps-api-1 | File "/usr/lib/python3.10/contextlib.py", line 135, in enter
wps-api-1 | return next(self.gen)
wps-api-1 | File "/app/app/auto_spatial_advisory/calculate_combustible_land_area.py", line 37, in get_fuel_types_from_object_store
wps-api-1 | fuel_types_layer = fuel_types.GetLayer()
wps-api-1 | AttributeError: 'NoneType' object has no attribute 'GetLayer'

**Reply by jmetev1** [reply]
Running upgrade 0669994d4089 -> d9c05cb16869, Add combustible area data to advisory_shapes

**Reply by conbrad** [reply]
> i'm on m1 pro so do you expect that i would have same issues or is it something specific to you?

It might be a resolved issue now since a lot of our dependencies hadn't built out support yet back when we had tried it, and we haven't tried it since then to my knowledge.

Regardless, related to the errors you're seeing now:
> i do need the alembic part right?

Yes alembic is our database migration tool, it'll need to run before the API connects to it. What it looks like is happening here is that migration requests data from our object store which requires credentials -- nothing wrong on your side, I'll need to figure out a solution for this. As a workaround you can comment out the `up` function of the body of the migration.

> i tried to move forward with the react upgrade but it's breaking the cypress tests and i can't really debug without a working server. one thing i considered is that i can disable auth on the frontend and point the frontend to the prod servers. but of course you have auth on api requests also. no way i can get around that easily eh? i assume it's not super private data.

Yes this is another tricky one. What you can do as a workaround is remove the auth locally, take a look at this diff to see what you can remove: https://github.com/bcgov/wps/compare/remove-hfi-local-auth?expand=1
